src/webgpu/api/operation/memory_sync/texture/texture_sync_test.ts

```diff
@@ -0,0 +1,144 @@
+  'command-buffer', // Operations are in different command buffers.
+  'pass', // Operations are in different passes.
+  'render-bundle', // Operations are in different render bundles.
+  'dispatch', // Operations are in different dispatches.
+  'draw', // Operations are in different draws.
+] as const;
+export type OperationBoundary = typeof kOperationBoundaries[number];
+
+// Context a particular operation is permitted in.
+export const kOperationContexts = [
+  'queue', // Operation occurs on the GPUQueue object
+  'command-encoder', // Operation may be encoded in a GPUCommandEncoder.
+  'compute-pass-encoder', // Operation may be encoded in a GPUComputePassEncoder.
+  'render-pass-encoder', // Operation may be encoded in a GPURenderPassEncoder.
+  'render-bundle-encoder', // Operation may be encoded in a GPURenderBundleEncoder.
+] as const;
+export type OperationContext = typeof kOperationContexts[number];
+
+interface BoundaryInfo {
+  readonly contexts: [OperationContext, OperationContext][];
+  // Add fields as needed
+}
+
+function combineContexts(
+  as: readonly OperationContext[],
+  bs: readonly OperationContext[]
+): [OperationContext, OperationContext][] {
+  const result: [OperationContext, OperationContext][] = [];
+  for (const a of as) {
+    for (const b of bs) {
+      result.push([a, b]);
+    }
+  }
+  return result;
+}
+
+const queueContexts = combineContexts(kOperationContexts, kOperationContexts);
+const commandBufferContexts = combineContexts(
+  kOperationContexts.filter(c => c !== 'queue'),
+  kOperationContexts.filter(c => c !== 'queue')
+);
+
+// Mapping of OperationBoundary => to a set of OperationContext pairs.
+// The boundary is capable of separating operations in those two contexts.
+export const kBoundaryInfo: {
+  readonly [k in OperationBoundary]: BoundaryInfo;
+} = /* prettier-ignore */ {
+  'queue': {
+    contexts: queueContexts,
+  },
+  'queue-op': {
+    contexts: queueContexts,
+  },
+  'command-buffer': {
+    contexts: commandBufferContexts,
+  },
+  'pass': {
+    contexts: [
+      ['compute-pass-encoder', 'compute-pass-encoder'],
+      ['compute-pass-encoder', 'render-pass-encoder'],
+      ['render-pass-encoder', 'compute-pass-encoder'],
+      ['render-pass-encoder', 'render-pass-encoder'],
+      ['render-bundle-encoder', 'render-pass-encoder'],
+      ['render-pass-encoder', 'render-bundle-encoder'],
+      ['render-bundle-encoder', 'render-bundle-encoder'],
+    ],
+  },
+  'render-bundle': {
+    contexts: [
+      ['render-bundle-encoder', 'render-pass-encoder'],
+      ['render-pass-encoder', 'render-bundle-encoder'],
+      ['render-bundle-encoder', 'render-bundle-encoder'],
+    ],
+  },
+  'dispatch': {
+    contexts: [
+      ['compute-pass-encoder', 'compute-pass-encoder'],
+    ],
+  },
+  'draw': {
+    contexts: [
+      ['render-pass-encoder', 'render-pass-encoder'],
+      ['render-bundle-encoder', 'render-pass-encoder'],
+    ],
+  },
+};
```

Comment:
I thought about this as well. Maybe it's worth it if I try out some of the combinations just to see the ergonomics.

Side note: One thing I thought about as well is changing the generation a bit so that we generate a bunch of tests - roughly one for each (context, boundary, context) triple, and then generate parameterizations for each of those.
It doesn't significantly help that it might be unwieldy to actually write the tests, but it does let us give each a separate description and .unimplemented() status should it be necessary.

Reply:
It would probably be good to sketch it out. Maybe an "ensure state" would be helpful generally? It probably wouldn't be terribly complex to write, although it might have to have a simplified worldview where render bundle is a state within render pass.

If using that helper then maybe instead of (context, boundary, context), the parameterization would look more like (context for first op, intermediate context, context for second op) which is almost the same, but with type (OperationContext, OperationContext, OperationContext) so it maps directly to three "ensure state" calls.

Reply:
> generate a bunch of tests - roughly one for each (context, boundary, context) triple

An alternative might be having a dictionary full of helper functions that perform the appropriate context switches but I think that would come out similarly to "ensure state".

Reply:
will sketch this out in a follow-up